tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/signatures/SetOperSignatures.scala

```diff
@@ -13,110 +13,90 @@ import at.forsyte.apalache.tla.typecomp.SignatureGenMap
-
-      val elemSetParis = ts.flatMap { tt =>
-        Seq(tt, SetT1(tt))
-      }
-      OperT1(t +: elemSetParis, SetT1(t))
-    }
+    val mapSig = mkSigMapEntry(map, mapPartial)
 
     // { x \in S: p } is polymorphic
     // (t, Set(t), Bool) => Set(t)
-    val filterSig = filter -> { _: Int =>
-      val t = varPool.fresh
-      OperT1(Seq(t, SetT1(t), BoolT1()), SetT1(t))
-    }
+    val filterSig = mkSigMapEntry(filter, { case Seq(t, st @ SetT1(tt), BoolT1()) if t == tt => st })
 
     // recSet does NOT have a signature (the return type depends on the field value)
 
     // Seq(S) is polymorphic, fixed arity 1
     // (Set(t)) => Set(Seq(t))
-    val seqSetSig = seqSet -> { _: Int =>
-      val t = varPool.fresh
-      OperT1(Seq(SetT1(t)), SetT1(SeqT1(t)))
-    }
+    val seqSetSig = mkSigMapEntry(seqSet, { case Seq(SetT1(t)) => SetT1(SeqT1(t)) })
 
     // SUBSET S is polymorphic, fixed arity 1
     // (Set(t)) => Set(Set(t))
-    val powSetSig = powerset -> { _: Int =>
-      val t = varPool.fresh
-      OperT1(Seq(SetT1(t)), SetT1(SetT1(t)))
+    val powSetSig = mkSigMapEntry(powerset, { case Seq(st: SetT1) => SetT1(st) })
+
+    val timesPartial: PartialSignature = {
+      case SetT1(t1) +: SetT1(t2) +: rest if rest.forall(_.isInstanceOf[SetT1]) =>
+        val ts = rest.map {
+          case SetT1(t) => t
+          case t        => t // impossible, but needed for warning suppression
+        }
+        SetT1(TupT1(t1 +: t2 +: ts: _*))
     }
 
     // \X is polyadic with n = 2 + k args
     // (Set(t1), ..., Set(tn)) => Set(<<t1,...,tn>>)
-    val timesSig = times -> { n: Int =>
-      val ts = varPool.fresh(n)
-      OperT1(ts.map(SetT1), SetT1(TupT1(ts: _*)))
-    }
+    val timesSig = mkSigMapEntry(times, timesPartial)
 
     // [_ -> _] is polymorphic
     // (Set(t1), Set(t2)) => Set(t1 -> t2)
-    val funSetSig = funSet -> { _: Int =>
-      val ts @ Seq(domT, cdmT) = varPool.fresh(2)
-      OperT1(ts.map(SetT1), SetT1(FunT1(domT, cdmT)))
-    }
+    val funSetSig = mkSigMapEntry(funSet, { case Seq(SetT1(t1), SetT1(t2)) => SetT1(FunT1(t1, t2)) })
 
     // UNION is polymorphic, fixed arity 1
     // (Set(Set(t))) => Set(t)
-    val unionSig = union -> { _: Int =>
-      val t = varPool.fresh
-      OperT1(Seq(SetT1(SetT1(t))), SetT1(t))
-    }
+    val unionSig = mkSigMapEntry(union, { case Seq(SetT1(st: SetT1)) => st })
 
     // { _ } is polyadic
     // (t, ..., t) => Set(t)
-    val enumSig = enumSet -> { n: Int =>
-      val t = varPool.fresh
-      OperT1(Seq.fill(n)(t), SetT1(t))
-    }
+    // We force a check against AnyPositiveArity, as the empty set is not built by an enumSet call.
+    // This should be caught by a require(_) in the builder method ahead of time, but having fallback
+    // sanity checking here helps with bugfixing.
+    val enumSig =
+      enumSet -> checkForArityException(enumSet.name, AnyPositiveArity(),
+          { case t +: ts if ts.forall(_ == t) => SetT1(t) })
```

Comment:
Shall we overload `mkSigMapEntry` to take an arity to have a uniform interface?

Reply:
Hmm, that one is explicitly intended to derive the operator arity, so I'd prefer not to. I'll consider adding an intermediate method (that gets called by `mkSigMapEntry` and calls `checkForArityException`) if any other builder method in the future batches needs this granularity, but for one-offs like this I'd prefer not to introduce more interface.

Reply:
Fair enough